Gateway API v1.5 blog post (#55360)
kubernetes/website: content/en/blog/_posts/2025/gateway-api-v1-5/index.md

---
layout: blog

> title: "Gateway API 1.5 new release"

Reviewer: ```suggestion
title: "Gateway API v1.5: Moving features to Stable"
```

Let's go for something a little more punchy?

Author: done

date: 2026-04-14T09:00:00-08:00
slug: gateway-api-v1-5
author: >
    [Katarzyna Łach](https://github.com/kl52752) (Google),
    [Nick Young](https://github.com/youngnick) (Isovalent),
    [Damian Sawicki](https://github.com/dsawicki) (Google),
    [Rostislav Bobrovsky](https://github.com/rostislavbobo) (Google),
    [Dave Protasowski,](https://github.com/dprotaso),
    [Gateway API Contributors](https://github.com/kubernetes-sigs/gateway-api/CHANGELOG/1.5-TEAM.md)
---

![Gateway API logo](gateway-api-logo.svg)

# Gateway API v1.5 Blogpost draft

> Gateway API has released v1.5.1, our biggest release yet!

Reviewer: This is on me, but on looking back at the v1.4 release blog, I think we can do better.

```suggestion
The Kubernetes SIG Network community presents the release of Gateway API (v1.5.1)! Released on March 14, 2025, version 1.5.1 is our biggest release yet, and concentrates on moving existing Experimental features to Standard (Stable).

Gateway API v1.5.0 brings five widely-requested feature promotions to the Standard channel (Gateway API's GA release channel):

ListenerSet
TLSRoute promoted to Stable
HTTPRoute CORS Filter
Certificate Selection for Gateway TLS Origination
ReferenceGrant promoted to Stable
```

Author: done

## New Release Process

As of Gateway API v1.5, the project has moved to a release train model, where on a feature freeze date, any features that are ready are shipped in the release.

This applies to both Experimental and Standard, and includes documentation.

> We are aiming for this to produce a more reliable release cadence (since we are basing our work off the excellent work done by SIG-Release on Kubernetes itself).

Reviewer: One more addition here:

```suggestion
We are aiming for this to produce a more reliable release cadence (since we are basing our work off the excellent work done by SIG-Release on Kubernetes itself).

As part of this change, we've also introduced Release Manager and Release Shadow roles to our release team. Many thanks to Flynn (Buoyant) and Beka Modebadze (Google) for all the great work coordinating and filing the rough edges of our release process. They are both going to continue in this role for the next release as well.
```

Author: done

---
layout: blog
title: "Gateway API v1.5: Moving features to Stable"
date: 2026-04-14T09:00:00-08:00
slug: gateway-api-v1-5
author: >
    [Katarzyna Łach](https://github.com/kl52752) (Google),
    [Nick Young](https://github.com/youngnick) (Isovalent),
    [Damian Sawicki](https://github.com/dsawicki) (Google),
    [Rostislav Bobrovsky](https://github.com/rostislavbobo) (Google),
    [Dave Protasowski,](https://github.com/dprotaso),

>     [Gateway API Contributors](https://github.com/kubernetes-sigs/gateway-api/CHANGELOG/1.5-TEAM.md)

Reviewer 2: This link is broken?

Author: done

---

![Gateway API logo](gateway-api-logo.svg)

# Gateway API v1.5: Moving features to Stable

The Kubernetes SIG Network community presents the release of Gateway API (v1.5)!

> Released on March 14, 2025, version 1.5 is our biggest release yet, and concentrates on moving existing Experimental features to Standard (Stable).

Reviewer 2: ```suggestion
Released on March 14, 2026, version 1.5 is our biggest release yet, and concentrates on moving existing Experimental features to Standard (Stable).
```

Author: done

Gateway API v1.5.0 brings five widely-requested feature promotions to the Standard channel (Gateway API's GA release channel):
* ListenerSet
* TLSRoute promoted to Stable
* HTTPRoute CORS Filter
* Certificate Selection for Gateway TLS Origination

> * ReferenceGrant promoted to Stable

Reviewer 2: There is some inconsistency here between versions: You say about 1.5.0, 1.5 and 1.5.1.

Maybe we should just stick with Gateway API v1.5 here, and then add a mention as a foot note that at the moment of writing this Blog post there was already Gateway API1.5.1 (or 1.5.2 depending on our timing) with some validation patches

Reviewer 3: I'd also vote for just "Gateway API v1.5" – usually the specific patch numbers aren't relevant for blog posts. Folks'll get the correct patch number when they go to install.

Author: changed

## New Release Process

As of Gateway API v1.5, the project has moved to a release train model, where on a feature freeze date, any features that are ready are shipped in the release.

> This applies to both Experimental and Standard, and includes documentation.

Reviewer 2: hum this phrase is confusing. What is the objective here? Linking the documentation? Otherwise I would drop this phrase

Reviewer 3: Maybe "...and also applies to documentation -- if the documentation isn't ready to ship, the feature isn't ready to ship." ?

Author: @youngnick can you help with this comment?

Reviewer: Yeah, I think that @kflynn's suggestion is right here. That's exactly what I was trying to say.

Author: done

> We are aiming for this to produce a more reliable release cadence (since we are basing our work off the excellent work done by SIG-Release on Kubernetes itself).

Reviewer 3: To be fair, it was more than _just_ SIG-Release, but I don't know if that's really worth noting. 😂

Author: @kflynn do you want some action from this comment ... :)

Reviewer 4: nit: it's actually "SIG Release" (no `-`) in our public comms

Author: done

As part of this change, we've also introduced Release Manager and Release Shadow roles to our release team. Many thanks to Flynn (Buoyant) and Beka Modebadze (Google) for all the great work coordinating and filing the rough edges of our release process. They are both going to continue in this role for the next release as well.

## New Standard features

### ListenerSet

Leads: [Dave Protasowski](https://github.com/dprotaso), [David Jumani](https://github.com/davidjumani)

[GEP-1713](https://gateway-api.sigs.k8s.io/geps/gep-1713/)

> #### Why ListenerSet?

Reviewer 2: maybe add a note about the need of Gateway to still have a listener today, as a limitation?

Author: sth like that:
```
While Gateway's Listener field remains broadly adopted and supported, the ListenerSet feature provides a viable alternative for specific implementations.
```

Reviewer 2: no no, I mean Listeners are not optional yet, so today it is a limitation that you still need to add listeners to a Gateway even if you just want to use ListenerSet

Author: Oh I misunderstood that. I added:
```
Even though the ListenerSet feature significantly enhances scalability, the Listener field in Gateway remains a mandatory requirement.
```
PTAL

Prior to `ListenerSet`, all listeners had to be specified directly on the `Gateway`. While this worked well for simple use cases, it created challenges for more complex or multi-tenant environments:

- Platform teams and application teams often needed to coordinate changes to the same `Gateway` resource
- Safely delegating ownership of individual listeners was difficult
- Extending existing Gateways required direct modification of the original resource

ListenerSet addresses these limitations by allowing listeners to be defined independently and then merged onto a target Gateway.

ListenerSets also enable attaching more than 64 listeners to a single, shared Gateway. This is critical for large scale deployments and scenarios with multiple hostnames per listener.

#### How it Works

> 

Reviewer 2: maybe add a link pointing to our docs guide?

Author: done

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: example-gateway
  namespace: infra
spec:
  gatewayClassName: example-gateway-class
  allowedListeners:
    namespaces:
      from: All # Selector let's you fine tune this
  listeners:
  - name: http
    protocol: HTTP
    port: 80
---
apiVersion: gateway.networking.k8s.io/v1
kind: ListenerSet
metadata:
  name: team-a-listeners
  namespace: team-a
spec:
  parentRef:
    name: example-gateway
    namespace: infra
  listeners:
  - name: https-a
    protocol: HTTPS
    port: 443
    hostname: a.example.com
    tls:
      certificateRefs:
      - name: a-cert
---
apiVersion: gateway.networking.k8s.io/v1
kind: ListenerSet
metadata:
  name: team-b-listeners
  namespace: team-b
spec:
  parentRef:
    name: example-gateway
    namespace: infra
  listeners:
  - name: https-b
    protocol: HTTPS
    port: 443
    hostname: b.example.com
    tls:
      certificateRefs:
      - name: b-cert
```

### TLSRoute

> Leads:   [Rostislav Bobrovsky](https://github.com/rostislavbobo), [Ricardo Pchevuzinske Katz](https://github.com/rikatz)

Reviewer 2: ```suggestion
Leads: [Rostislav Bobrovsky](https://github.com/rostislavbobo), [Ricardo Pchevuzinske Katz](https://github.com/rikatz)
```

Author: done

[GEP-2643](https://gateway-api.sigs.k8s.io/geps/gep-2643/)

> The [TLSRoute resource](https://gateway-api.sigs.k8s.io/api-types/tlsroute/) allows you to multiplex TLS connections by matching the Server Name Indication (SNI) presented by the client during the TLS handshake and directing the stream to the appropriate Kubernetes backends.

Reviewer 2: multiplex seems a difficult word :D 

maybe "allows you to route requests by matching the Server Name Indication (SNI)" and also mention this is also known as passthrough?

Author: done

> The [TLSRoute resource](https://gateway-api.sigs.k8s.io/api-types/tlsroute/) allows you to allows you to route requests by matching the Server Name Indication (SNI) presented by the client during the TLS handshake and directing the stream to the appropriate Kubernetes backends.

Reviewer 2: ```suggestion
The [TLSRoute resource](https://gateway-api.sigs.k8s.io/api-types/tlsroute/) allows you to route requests by matching the Server Name Indication (SNI) presented by the client during the TLS handshake and directing the stream to the appropriate Kubernetes backends.
```

Author: done

When working with `TLSRoute`, a `Gateway` TLS listener can be configured in one of two modes: **`Passthrough`** or **`Terminate`**.

> **If you install v1.5.1 Standard over v1.4 or earlier Experimental, your existing Experimental TLSRoutes will not be usable**. This is because they will be stored in the `v1alpha2` or `v1alpha3` version, which is _not_ included in the v1.5.1 Standard YAMLs. If this applies to you, either continue using Experimental for v1.5.1 and onward, or you'll need to download and migrate your TLSRoutes to `v1`, which _is_ present in the Standard YAMLs.

Reviewer 2: ```suggestion
**If you install v1.5 Standard over v1.4 or earlier Experimental, your existing Experimental TLSRoutes will not be usable**. This is because they will be stored in the `v1alpha2` or `v1alpha3` version, which is _not_ included in the v1.5 Standard YAMLs. If this applies to you, either continue using Experimental for v1.5 and onward, or you'll need to download and migrate your TLSRoutes to `v1`, which _is_ present in the Standard YAMLs.
```

Author: done

#### Passthrough Mode

> The `Passthrough` mode is designed for strict security requirements. It is ideal for scenarios where traffic must remain encrypted end-to-end until it reaches the destination backend, when the external client and backend need to authenticate directly with each other, or when you can’t store certificates on the `Gateway`.

Reviewer 2: it is also valid for cases where you need an encrypted TCP stream, and not a HTTP traffic. Probably not worth calling it out on the docs, but I though would be good to raise it

Author: added

    - name: foo-svc
      port: 8443
```

#### Terminate Mode
The `Terminate` mode provides the convenience of centralized TLS certificate management directly at the `Gateway`.

In this mode, the TLS session is fully terminated at the `Gateway`, which then routes the decrypted payload to the destination backend as a raw TCP stream.

The following `TLSRoute` is attached to a listener that is configured in `Terminate` mode. It will match only TLS handshakes with the `bar.example.com` SNI hostname and apply its routing rules to pass the decrypted TCP stream to the configured backend:
```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: example-gateway
spec:
  gatewayClassName: example-gateway-class
  listeners:
  - name: tls-terminate
    protocol: TLS
    port: 443
    tls:
      mode: Terminate
      certificateRefs:
      - name: tls-terminate-certificate
---
apiVersion: gateway.networking.k8s.io/v1
kind: TLSRoute
metadata:
  name: bar-route
spec:
  parentRefs:
  - name: example-gateway
    sectionName: tls-terminate
  hostnames:
  - "bar.example.com"
  rules:
  - backendRefs:
    - name: bar-svc
      port: 8080
```

### HTTPRoute CORS filter

> Leads:   [Damian Sawicki](https://github.com/DamianSawicki), [Ricardo Pchevuzinske Katz](https://github.com/rikatz), [Norwin Schnyder](https://github.com/snorwin), [Huabing (Robin) Zhao](https://github.com/zhaohuabing), [LiangLliu](https://github.com/LiangLliu),

Reviewer 2: ```suggestion
Leads: [Damian Sawicki](https://github.com/DamianSawicki), [Ricardo Pchevuzinske Katz](https://github.com/rikatz), [Norwin Schnyder](https://github.com/snorwin), [Huabing (Robin) Zhao](https://github.com/zhaohuabing), [LiangLliu](https://github.com/LiangLliu),
```

btw I am suggesting more the removal of spaces here to keep consistency with the first item (1 vs 3 spaces) but feel free to ignore me!

Author: done

[GEP-1767](https://gateway-api.sigs.k8s.io/geps/gep-1767/)

> Cross-origin resource sharing (CORS) is an HTTP-header based security mechanism that allows (or denies) a web page to access resources from a server on an origin different from the domain that served the web page. See our [documentation page](../guides/http-cors) for more information.

Reviewer 2: link here may also be broken

Author: done

> The [HTTPRoute resource](../api-types/httproute.md) can be used to configure Cross-Origin Resource Sharing (CORS). The following `HTTPRoute` allows requests from `https://app.example`:

Reviewer 2: this link will be broken on Kubernetes main website, probably needs fixing

Author: done

> ```yaml=

Reviewer 2: nit: keep the codeblock marker consistency (eg yaml= vs yaml)

Author: done

apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: cors
spec:
  parentRefs:
  - name: same-namespace
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /cors-behavior-creds-false
    backendRefs:
    - name: infra-backend-v1
      port: 8080
    filters:
    - cors:
        allowOrigins:
        - https://app.example
      type: CORS

> 

Reviewer 2: nit: remove the last line

Author: done

```

Instead of specifying a list of specific origins, you can also specify a single wildcard (`"*"`), which will allow any origin. It is also allowed to use semi-specified origins in the list, where the wildcard appears after the scheme and at the beginning of the hostname, e.g. `https://*.bar.com`:

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: cors
spec:
  parentRefs:
  - name: same-namespace
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /cors-behavior-creds-false
    backendRefs:
    - name: infra-backend-v1
      port: 8080
    filters:
    - cors:
        allowOrigins:
        - https://www.baz.com
        - https://*.bar.com
        - https://*.foo.com
      type: CORS
```
The `CORS` filter also allows you to specify other CORS options, such as:

> - `allowCredentials`: Specifies whether the browser is allowed to include credentials (such as cookies and HTTP authentication) in the CORS request.

Reviewer 2: my personal thinking: I think that we could be a bit more explicit here, given CORS is a huge confusing pain :P 

Maybe say what headers should be set by the proxy when these fields are set?

Author: changed: 
- "The `CORS` filter also allows you to specify other CORS options, such as"
+ "HTTPRoute filters allow for the configuration of `CORS` settings. See a list of supported options below:"

PTAL

2.  **Per-Port Configuration**: This allows for fine-grained control, overriding the default configuration for all listeners handling traffic on a specific port.

Example:

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: client-validation-basic
spec:
  gatewayClassName: acme-lb
  tls:
    frontend:
      default:
        validation:
          caCertificateRefs:
          - kind: ConfigMap
            group: ""
            name: foo-example-com-ca-cert
      perPort:
      - port: 8443
        tls:
          validation:
            caCertificateRefs:
            - kind: ConfigMap
              group: ""
              name: foo-example-com-ca-cert
            mode: "AllowInsecureFallback"
  listeners:
  - name: foo-https
    protocol: HTTPS
    port: 443
    hostname: foo.example.com
    tls:
      certificateRefs:
      - kind: Secret
        group: ""
        name: foo-example-com-cert
  - name: bar-https
    protocol: HTTPS
    port: 8443
    hostname: bar.example.com
    tls:
      certificateRefs:
      - kind: Secret
        group: ""
        name: bar-example-com-cert
```

### Certificate selection for Gateway TLS origination

> Leads: [Marcin Kosieradzki](https://github.com/mkosieradzki), [Rob Scott](https://github.com/robscott),  [Norwin Schnyder](https://github.com/snorwin), [Lior Lieberman](https://github.com/LiorLieberman), [Katarzyna Lach](https://github.com/kl52752)

Reviewer 2: ```suggestion
Leads: [Marcin Kosieradzki](https://github.com/mkosieradzki), [Rob Scott](https://github.com/robscott), [Norwin Schnyder](https://github.com/snorwin), [Lior Lieberman](https://github.com/LiorLieberman), [Katarzyna Lach](https://github.com/kl52752)
```

Author: done

[GEP-3155](https://gateway-api.sigs.k8s.io/geps/gep-3155/)

Mutual TLS (mTLS) for upstream connections requires the Gateway to present a client certificate to the backend, in addition to verifying the backend's certificate. This ensures that the backend only accepts connections from authorized Gateways.

#### Gateway’s Client Certificate Configuration
To configure the client certificate that the Gateway uses when connecting to backends, use the `tls.backend.clientCertificateRef` field in the `Gateway` resource.
This configuration applies to the Gateway as a client for *all* upstream connections managed by that Gateway.

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: backend-tls
spec:
  gatewayClassName: acme-lb
  tls:
    backend:
      clientCertificateRef:
        kind: Secret
        group: ""
        name: foo-example-cert
  listeners:
  - name: foo-http
    protocol: HTTP
    port: 80
    hostname: foo.example.com
```

### ReferenceGrant promoted to v1

> The ReferenceGrant resource has not changed in more than a year, and we do not expect it to change further, so its version has been bumped to `v1`, and it is now officially in the Standard channel, and abides by the GA API contract (that is, no breaking changes).

Reviewer: @kflynn is right, we need some ending text. Here's my suggestion, based on the previous blog posts:

```suggestion
The ReferenceGrant resource has not changed in more than a year, and we do not expect it to change further, so its version has been bumped to `v1`, and it is now officially in the Standard channel, and abides by the GA API contract (that is, no breaking changes).

## Try it out

Unlike other Kubernetes APIs, you don't need to upgrade to the latest version of
Kubernetes to get the latest version of Gateway API. As long as you're running
Kubernetes 1.30 or later, you'll be able to get up and running with this version
of Gateway API.

To try out the API, follow the [Getting Started Guide](https://gateway-api.sigs.k8s.io/guides/).

As of this writing, seven implementations are already fully conformant with Gateway API v1.5. In alphabetical order:

- [NGINX Gateway Fabric](https://github.com/nginx/nginx-gateway-fabric/releases/tag/v2.5.0)
- [Airlock Microgateway](https://github.com/airlock/microgateway/releases/tag/5.0.0)
- [GKE Gateway](https://docs.cloud.google.com/kubernetes-engine/docs/concepts/gateway-api)
- [HAProxy Ingress](https://github.com/jcmoraisjr/haproxy-ingress/releases/tag/v0.17.0-alpha.1)
- [kgateway](https://github.com/kgateway-dev/kgateway/releases/tag/v2.3.0-beta.3)
- [Traefik Proxy](https://github.com/traefik/traefik/releases/tag/v3.7.0-rc.1)

## Get involved

Wondering when a feature will be added?  There are lots of opportunities to get
involved and help define the future of Kubernetes routing APIs for both ingress
and service mesh.

* Check out the [user guides](https://gateway-api.sigs.k8s.io/guides) to see what use-cases can be addressed.
* Try out one of the [existing Gateway controllers](https://gateway-api.sigs.k8s.io/implementations/).
* Or [join us in the community](https://gateway-api.sigs.k8s.io/contributing/)
and help us build the future of Gateway API together!

The maintainers would like to thank _everyone_ who's contributed to Gateway
API, whether in the form of commits to the repo, discussion, ideas, or general
support. We could never have made this kind of progress without the support of
this dedicated and active community.
```

Author: added

---
layout: blog
title: "Gateway API v1.5: Moving features to Stable"

> date: 2026-04-14T09:00:00-08:00

Reviewer 4: I promise you that this is right; all blog articles initially merge as a draft.
```suggestion
draft: true
```

Author: done

slug: gateway-api-v1-5
authors: >
    [Katarzyna Łach](https://github.com/kl52752) (Google),
    [Nick Young](https://github.com/youngnick) (Isovalent),
    [Damian Sawicki](https://github.com/dsawicki) (Google),
    [Rostislav Bobrovsky](https://github.com/rostislavbobo) (Google),
    [Dave Protasowski](https://github.com/dprotaso),
---

![Gateway API logo](gateway-api-logo.svg)

> # Gateway API v1.5: Moving features to Stable

Reviewer 4: ```suggestion
```

Author: done

The Kubernetes SIG Network community presents the release of Gateway API (v1.5)!
Released on March 14, 2026, version 1.5 is our biggest release yet, and concentrates on moving existing Experimental features to Standard (Stable).

> >Gateway API [v1.5.1](https://github.com/kubernetes-sigs/gateway-api/releases/tag/v1.5.1) patch release is already available

Reviewer 4: ```suggestion

The Gateway API [v1.5.1](https://github.com/kubernetes-sigs/gateway-api/releases/tag/v1.5.1) patch release is already available.
```

(don't mark it as a quotation)

Author: done

Gateway API v1.5 brings five widely-requested feature promotions to the Standard channel (Gateway API's GA release channel):
* ListenerSet
* TLSRoute promoted to Stable
* HTTPRoute CORS Filter

> * Certificate Selection for Gateway TLS Origination

Reviewer 2: did you missed "### Gateway Client Certificate validation" ? :D

Author: thanks for noticed that

* ReferenceGrant promoted to Stable

Special thanks for [Gateway API Contributors](https://github.com/kubernetes-sigs/gateway-api/blob/a811d174a406553006bbb9a3594b49380cb9069e/CHANGELOG/1.5-TEAM.md) for their efforts on this release.

> ## New Release Process

Reviewer 4: ```suggestion
## New release process
```

Author: done

As of Gateway API v1.5, the project has moved to a release train model, where on a feature freeze date, any features that are ready are shipped in the release.

This applies to both Experimental and Standard, and also applies to documentation -- if the documentation isn't ready to ship, the feature isn't ready to ship.

We are aiming for this to produce a more reliable release cadence (since we are basing our work off the excellent work done by SIG-Release on Kubernetes itself).
As part of this change, we've also introduced Release Manager and Release Shadow roles to our release team. Many thanks to Flynn (Buoyant) and Beka Modebadze (Google) for all the great work coordinating and filing the rough edges of our release process. They are both going to continue in this role for the next release as well.

## New Standard features

### ListenerSet

Leads: [Dave Protasowski](https://github.com/dprotaso), [David Jumani](https://github.com/davidjumani)

[GEP-1713](https://gateway-api.sigs.k8s.io/geps/gep-1713/)

#### Why ListenerSet?

> Prior to ListenerSet`, all listeners had to be specified directly on the `Gateway`. While this worked well for simple use cases, it created challenges for more complex or multi-tenant environments:

Reviewer 2: I think this "`¨ will break and make markdown understanding this as codeformatting (need to doublecheck but netlify is down)

Author: yeah this might be it, thanks!

> Prior to `ListenerSet`, all listeners had to be specified directly on the `Gateway`. While this worked well for simple use cases, it created challenges for more complex or multi-tenant environments:

Reviewer 4: For this website we use the Kubernetes project docs style guide, so:
```suggestion
Prior to ListenerSet, all listeners had to be specified directly on the Gateway object.
While this worked well for simple use cases, it created challenges for more complex
or multi-tenant environments:
```

Author: done

> - Platform teams and application teams often needed to coordinate changes to the same `Gateway` resource

Reviewer 4: ```suggestion
- Platform teams and application teams often needed to coordinate changes to the same Gateway
```
(or "same Gateway object"; Gateways are resources but we also know that they are objects)

Author: done

- Safely delegating ownership of individual listeners was difficult
- Extending existing Gateways required direct modification of the original resource

[ListenerSet](https://gateway-api.sigs.k8s.io/guides/listener-set/) addresses these limitations by allowing listeners to be defined independently and then merged onto a target Gateway.

ListenerSets also enable attaching more than 64 listeners to a single, shared Gateway. This is critical for large scale deployments and scenarios with multiple hostnames per listener.

> > Even though the ListenerSet feature significantly enhances scalability, the Listener field in Gateway remains a mandatory requirement.

Reviewer 4: ```suggestion
Even though the ListenerSet feature significantly enhances scalability, the `listener` field in Gateway
**remains a mandatory requirement**.
```

Author: done

> #### How it Works

Reviewer 4: ```suggestion
#### How it works
```

Author: done

> A `ListenerSet` attaches to a `Gateway` and contributes one or more listeners. The Gateway controller is responsible for merging listeners from the `Gateway` resource itself and any attached `ListenerSet` resources.

Reviewer 4: ```suggestion
A ListenerSet attaches to a Gateway and contributes one or more listeners.
The gateway controller is responsible for merging listeners from the Gateway resource itself and any attached ListenerSets.
```

Author: done

> In this example, a central infrastructure team defines a `Gateway` with a default HTTP listener, while two different application teams define their own `ListenerSet` resources in separate namespaces. Both ListenerSets attach to the same Gateway and contribute additional HTTPS listeners.

Reviewer 4: ```suggestion
In this example, a central infrastructure team defines a Gateway with a default HTTP listener,
while two different application teams define their own ListenerSet resources in separate namespaces.
Both ListenerSets attach to the same Gateway and contribute additional HTTPS listeners.

```

Author: done

```yaml

> apiVersion: gateway.networking.k8s.io/v1

Reviewer 4: (nit)
```suggestion
---
apiVersion: gateway.networking.k8s.io/v1
```

Author: done

kind: Gateway
metadata:
  name: example-gateway
  namespace: infra
spec:
  gatewayClassName: example-gateway-class
  allowedListeners:
    namespaces:

>       from: All # Selector let's you fine tune this

Reviewer 4: ```suggestion
      from: All # A selector lets you fine tune this
```

Author: done

  listeners:
  - name: http
    protocol: HTTP
    port: 80
---
apiVersion: gateway.networking.k8s.io/v1
kind: ListenerSet
metadata:
  name: team-a-listeners
  namespace: team-a
spec:
  parentRef:
    name: example-gateway
    namespace: infra
  listeners:
  - name: https-a
    protocol: HTTPS
    port: 443
    hostname: a.example.com
    tls:
      certificateRefs:
      - name: a-cert
---
apiVersion: gateway.networking.k8s.io/v1
kind: ListenerSet
metadata:
  name: team-b-listeners
  namespace: team-b
spec:
  parentRef:
    name: example-gateway
    namespace: infra
  listeners:
  - name: https-b
    protocol: HTTPS
    port: 443
    hostname: b.example.com
    tls:
      certificateRefs:
      - name: b-cert
```

### TLSRoute

> Leads: [Rostislav Bobrovsky](https://github.com/rostislavbobo), [Ricardo Pchevuzinske Katz](https://github.com/rikatz)

Reviewer 4: (nit)
```suggestion

Leads: [Rostislav Bobrovsky](https://github.com/rostislavbobo), [Ricardo Pchevuzinske Katz](https://github.com/rikatz)
```

Author: done

[GEP-2643](https://gateway-api.sigs.k8s.io/geps/gep-2643/)

The [TLSRoute resource](https://gateway-api.sigs.k8s.io/api-types/tlsroute/) allows you to route requests by matching the Server Name Indication (SNI) presented by the client during the TLS handshake and directing the stream to the appropriate Kubernetes backends.

> When working with `TLSRoute`, a `Gateway` TLS listener can be configured in one of two modes: **`Passthrough`** or **`Terminate`**.

Reviewer 4: ```suggestion
When working with TLSRoute, a Gateway's TLS listener can be configured in one of two modes: `Passthrough` or `Terminate`.
```

Author: done

> **If you install v1.5 Standard over v1.4 or earlier Experimental, your existing Experimental TLSRoutes will not be usable**. This is because they will be stored in the `v1alpha2` or `v1alpha3` version, which is _not_ included in the v1.5 Standard YAMLs. If this applies to you, either continue using Experimental for v1.5.1 and onward, or you'll need to download and migrate your TLSRoutes to `v1`, which _is_ present in the Standard YAMLs.

Reviewer 4: ```suggestion
**If you install Gateway API v1.5 Standard over v1.4 or earlier Experimental, your existing
Experimental TLSRoutes will not be usable**.
This is because they will be stored in the `v1alpha2` or `v1alpha3` version, which is **_not_** included in the v1.5 Standard manifests for Gateway API.
If this applies to you, either continue using Experimental for v1.5.1 and onward, or you'll need to download and migrate your TLSRoutes to `v1`, which _is_ present in the Standard manifests.
```

Author: done

> #### Passthrough Mode

Reviewer 4: ```suggestion
#### Passthrough mode
```

Author: done

The `Passthrough` mode is designed for strict security requirements. It is ideal for scenarios where traffic must remain encrypted end-to-end until it reaches the destination backend, when the external client and backend need to authenticate directly with each other, or when you can’t store certificates on the `Gateway`. This configuration is also applicable when an encrypted TCP stream is required instead of standard HTTP traffic.

> In this mode, the encrypted byte stream is proxied directly to the destination backend. The `Gateway` has zero access to private keys or unencrypted data.

Reviewer 4: ```suggestion
In this mode, the encrypted byte stream is proxied directly to the destination backend. The Gateway has zero access to private keys or unencrypted data.
```

Author: done

> The following `TLSRoute` is attached to a listener that is configured in `Passthrough` mode. It will match only TLS handshakes with the `foo.example.com` SNI hostname and apply its routing rules to pass the encrypted TCP stream to the configured backend:

Reviewer 4: ```suggestion
The following TLSRoute is attached to a listener that is configured in `Passthrough` mode. It will match only TLS handshakes with the `foo.example.com` SNI hostname and apply its routing rules to pass the encrypted TCP stream to the configured backend:

```

Reviewer 4: nit: we could comment about compatibility with encrypted SNI

```yaml

> apiVersion: gateway.networking.k8s.io/v1

Reviewer 4: (nit)
```suggestion
---
apiVersion: gateway.networking.k8s.io/v1
```

Author: done

kind: Gateway
metadata:
  name: example-gateway
spec:
  gatewayClassName: example-gateway-class
  listeners:
  - name: tls-passthrough
    protocol: TLS
    port: 8443
    tls:
      mode: Passthrough
---
apiVersion: gateway.networking.k8s.io/v1
kind: TLSRoute
metadata:
  name: foo-route
spec:
  parentRefs:
  - name: example-gateway
    sectionName: tls-passthrough
  hostnames:
  - "foo.example.com"
  rules:
  - backendRefs:
    - name: foo-svc
      port: 8443
```

> #### Terminate Mode

Reviewer 4: ```suggestion
#### Terminate mode

```

Author: done

> The `Terminate` mode provides the convenience of centralized TLS certificate management directly at the `Gateway`.

Reviewer 4: ```suggestion

The `Terminate` mode provides the convenience of centralized TLS certificate management directly at the Gateway.
```

Author: done

> In this mode, the TLS session is fully terminated at the `Gateway`, which then routes the decrypted payload to the destination backend as a raw TCP stream.

Reviewer 4: ```suggestion
In this mode, the TLS session is fully terminated at the Gateway, which then
routes the decrypted payload to the destination backend as a plain text TCP stream.
```
"raw" implies "not wrapped in HTTP", and I don't think that's right.

Author: done

> The following `TLSRoute` is attached to a listener that is configured in `Terminate` mode. It will match only TLS handshakes with the `bar.example.com` SNI hostname and apply its routing rules to pass the decrypted TCP stream to the configured backend:

Reviewer 4: ```suggestion
The following TLSRoute is attached to a listener that is configured in `Terminate` mode. 
It will match only TLS handshakes with the `bar.example.com` SNI hostname and apply its routing rules to pass the decrypted TCP stream to the configured backend:

```

Author: done

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: example-gateway
spec:
  gatewayClassName: example-gateway-class
  listeners:
  - name: tls-terminate
    protocol: TLS
    port: 443
    tls:
      mode: Terminate
      certificateRefs:
      - name: tls-terminate-certificate
---
apiVersion: gateway.networking.k8s.io/v1
kind: TLSRoute
metadata:
  name: bar-route
spec:
  parentRefs:
  - name: example-gateway
    sectionName: tls-terminate
  hostnames:
  - "bar.example.com"
  rules:
  - backendRefs:
    - name: bar-svc
      port: 8080
```

> ### HTTPRoute CORS filter

Reviewer 4: ```suggestion
### HTTPRoute CORS filter

```

Author: done

Leads: [Damian Sawicki](https://github.com/DamianSawicki), [Ricardo Pchevuzinske Katz](https://github.com/rikatz), [Norwin Schnyder](https://github.com/snorwin), [Huabing (Robin) Zhao](https://github.com/zhaohuabing), [LiangLliu](https://github.com/LiangLliu),

[GEP-1767](https://gateway-api.sigs.k8s.io/geps/gep-1767/)

Cross-origin resource sharing (CORS) is an HTTP-header based security mechanism that allows (or denies) a web page to access resources from a server on an origin different from the domain that served the web page. See our [documentation page](https://gateway-api.sigs.k8s.io/guides/http-cors/) for more information.

> The [HTTPRoute resource](https://gateway-api.sigs.k8s.io/api-types/httproute/) can be used to configure Cross-Origin Resource Sharing (CORS). The following `HTTPRoute` allows requests from `https://app.example`:

Reviewer 4: ```suggestion
The [HTTPRoute resource](https://gateway-api.sigs.k8s.io/api-types/httproute/) can be used to configure Cross-Origin Resource Sharing (CORS). The following HTTPRoute allows requests from `https://app.example`:
```

Author: done

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: cors
spec:
  parentRefs:
  - name: same-namespace
  rules:
  - matches:
    - path:
      type: PathPrefix
      value: /cors-behavior-creds-false
    backendRefs:
    - name: infra-backend-v1
      port: 8080
    filters:
    - cors:
      allowOrigins:
        - https://www.baz.com
        - https://*.bar.com
        - https://*.foo.com
      type: CORS
```
HTTPRoute filters allow for the configuration of `CORS` settings. See a list of supported options below:

- `allowCredentials`: Specifies whether the browser is allowed to include credentials (such as cookies and HTTP authentication) in the CORS request.
- `allowMethods`: The HTTP methods that are allowed for CORS requests.
- `allowHeaders`: The HTTP headers that are allowed for CORS requests.
- `exposeHeaders`: The HTTP headers that are exposed to the client.
- `maxAge`: The maximum time in seconds that the browser should cache the preflight response.

> 

Reviewer 4: Even better: use a Markdown description list, eg
```markdown
`allowCredentials`
: Specifies whether the browser is allowed to include credentials (such as cookies and HTTP authentication) in the CORS request.

```
(only the first item is shown)

A comprehensive example:
```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: cors-allow-credentials
spec:
  parentRefs:
  - name: same-namespace
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /cors-behavior-creds-true
    backendRefs:
    - name: infra-backend-v1
      port: 8080
    filters:
    - cors:
        allowOrigins:

>         - "https://www.foo.com"

Reviewer 4: Use example domains, eg `foo.example.com` or `foo
example`

Author: done

        - "https://*.bar.com"
        allowMethods:
        - GET
        - OPTIONS
        allowHeaders:
        - "*"
        exposeHeaders:
        - "x-header-3"
        - "x-header-4"
        allowCredentials: true
        maxAge: 3600
      type: CORS
```

> ### Gateway Client Certificate validation

Reviewer 4: ```suggestion
### Gateway client certificate validation

```

Author: done

Leads: [Arko Dasgupta](https://github.com/arkodg), [Katarzyna Łach](https://github.com/kl52752), [Norwin Schnyder](https://github.com/snorwin)

[GEP-91](https://gateway-api.sigs.k8s.io/geps/gep-91/)

Client certificate validation, also known as mutual TLS (mTLS), is a security mechanism where the client provides a certificate to the server to prove its identity. This is in contrast to standard TLS, where only the server presents a certificate to the client.
In the context of the Gateway API, frontend mTLS means that the Gateway validates the client's certificate before allowing the connection to proceed to a backend service. This validation is done by checking the client certificate against a set of trusted Certificate Authorities (CAs) configured on the Gateway. The API was shaped this way to address a critical security vulnerability related to connection reuse and still provide some level of flexibility.

> #### Configuration Overview

Reviewer 4: ```suggestion
#### Configuration overview

```
Even better: don't include this here; instead, hyperlink to the relevant documentation.

Author: done

Client validation is defined using the `frontendValidation` struct, which specifies how the Gateway should verify the client's identity.

> *   **`caCertificateRefs`**: A list of references to Kubernetes objects (typically `ConfigMap`s) containing PEM-encoded CA certificate bundles used as trust anchors to validate the client's certificate.

Reviewer 4: (nit)
Even better: use a Markdown description list here.

            caCertificateRefs:
            - kind: ConfigMap
              group: ""
              name: foo-example-com-ca-cert
            mode: "AllowInsecureFallback"
  listeners:
  - name: foo-https
    protocol: HTTPS
    port: 443
    hostname: foo.example.com
    tls:
      certificateRefs:
      - kind: Secret
        group: ""
        name: foo-example-com-cert
  - name: bar-https
    protocol: HTTPS
    port: 8443
    hostname: bar.example.com
    tls:
      certificateRefs:
      - kind: Secret
        group: ""
        name: bar-example-com-cert
```

### Certificate selection for Gateway TLS origination
Leads: [Marcin Kosieradzki](https://github.com/mkosieradzki), [Rob Scott](https://github.com/robscott), [Norwin Schnyder](https://github.com/snorwin), [Lior Lieberman](https://github.com/LiorLieberman), [Katarzyna Lach](https://github.com/kl52752)

[GEP-3155](https://gateway-api.sigs.k8s.io/geps/gep-3155/)

Mutual TLS (mTLS) for upstream connections requires the Gateway to present a client certificate to the backend, in addition to verifying the backend's certificate. This ensures that the backend only accepts connections from authorized Gateways.

> #### Gateway’s Client Certificate Configuration

Reviewer 4: ```suggestion
#### Gateway’s client certificate configuration

```

Author: done

> To configure the client certificate that the Gateway uses when connecting to backends, use the `tls.backend.clientCertificateRef` field in the `Gateway` resource.

Reviewer 4: ```suggestion

To configure the client certificate that the Gateway uses when connecting to backends,
use the `tls.backend.clientCertificateRef` field in the Gateway object.
```

Author: done

> This configuration applies to the Gateway as a client for *all* upstream connections managed by that Gateway.

Reviewer 4: ```suggestion
This configuration applies to the Gateway as a client for **all** upstream connections managed by that Gateway.
```

Author: done

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: backend-tls
spec:
  gatewayClassName: acme-lb
  tls:
    backend:
      clientCertificateRef:
        kind: Secret

>         group: ""

Reviewer 4: (nit)
```suggestion
        group: "" # empty string means core API group
```

Author: done

As of this writing, seven implementations are already fully conformant with Gateway API v1.5. In alphabetical order:
- [NGINX Gateway Fabric](https://github.com/nginx/nginx-gateway-fabric/releases/tag/v2.5.0)
- [Airlock Microgateway](https://github.com/airlock/microgateway/releases/tag/5.0.0)
- [GKE Gateway](https://docs.cloud.google.com/kubernetes-engine/docs/concepts/gateway-api)
- [HAProxy Ingress](https://github.com/jcmoraisjr/haproxy-ingress/releases/tag/v0.17.0-alpha.1)
- [kgateway](https://github.com/kgateway-dev/kgateway/releases/tag/v2.3.0-beta.3)
- [Traefik Proxy](https://github.com/traefik/traefik/releases/tag/v3.7.0-rc.1)

## Get involved

Wondering when a feature will be added?  There are lots of opportunities to get
involved and help define the future of Kubernetes routing APIs for both ingress
and service mesh.
* Check out the [user guides](https://gateway-api.sigs.k8s.io/guides) to see what use-cases can be addressed.
* Try out one of the [existing Gateway controllers](https://gateway-api.sigs.k8s.io/implementations/).
* Or [join us in the community](https://gateway-api.sigs.k8s.io/contributing/) and help us build the future of Gateway API together!

> The maintainers would like to thank _everyone_ who's contributed to Gateway

Reviewer 4: nit: use bold, not italics, here

Author: done

---
layout: blog
title: "Gateway API v1.5: Moving features to Stable"
draft: true
slug: gateway-api-v1-5
authors: >
    [Katarzyna Łach](https://github.com/kl52752) (Google),
    [Nick Young](https://github.com/youngnick) (Isovalent),
    [Damian Sawicki](https://github.com/dsawicki) (Google),
    [Rostislav Bobrovsky](https://github.com/rostislavbobo) (Google),
    [Dave Protasowski](https://github.com/dprotaso),
---

![Gateway API logo](gateway-api-logo.svg)

The Kubernetes SIG Network community presents the release of Gateway API (v1.5)!
Released on March 14, 2026, version 1.5 is our biggest release yet, and concentrates on moving existing Experimental features to Standard (Stable).

The Gateway API [v1.5.1](https://github.com/kubernetes-sigs/gateway-api/releases/tag/v1.5.1) patch release is already available

The Gateway API v1.5 brings five widely-requested feature promotions to the Standard channel (Gateway API's GA release channel):
* ListenerSet
* TLSRoute promoted to Stable
* HTTPRoute CORS Filter
* Certificate Selection for Gateway TLS Origination

> * ReferenceGrant promoted to Stable

Reviewer 2: still missing "Gateway Client Certificate validation" on this list (and change the phrasing above from "five" to "six" widely-requested...)
